File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        d = {"(":")", "{":"}", "[":"]"}
        
        for i in s:
            if i in d:
                stack.append(i)
            else:
                if not stack or d[stack[-1]]!=i:
                    return(False)
                stack.pop()
        
        return(True if not stack else False)

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> d = new HashMap<>();
        d.put('[',']');
        d.put('{','}');
        d.put('(',')');
        Stack<Character> myStack = new Stack<>();
        for (int i=0; i<s.length(); i++){
            char presentChar = s.charAt(i);
            if(d.containsKey(presentChar))
                myStack.push(presentChar);
            else if(!myStack.isEmpty() && d.get(myStack.peek()) == presentChar) 
                myStack.pop();
            else 
                return false;
        }

        return myStack.isEmpty();
        }
    }